### PR TITLE
Add share retro functionality via magic link

### PR DESCRIPTION
**Pull request by othmankh**
* A short explanation of the proposed change:
Add share retro functionality to be able to join any retro even protected ones with a magic link.

* An explanation of the use cases your change solves
Participants no longer need to know or remember the password in order to join retros.

* Links to any other associated PRs

* [x] I have reviewed the [contributing guide](https://github.com/pivotal/postfacto/blob/master/CONTRIBUTING.md)

* [x] I have made this pull request to the `master` branch

* [x] I have run all the tests using `./test.sh`.

* [x] I have added the [copyright headers](https://github.com/pivotal/postfacto/blob/master/license-header.txt) to each new file added

* [x] I have given myself credit in the [humans.txt](https://github.com/pivotal/postfacto/blob/master/humans.txt) file (assuming I want to)

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@othmankh Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=pivotal/postfacto&pullRequestId=288)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=pivotal/postfacto&pullRequestId=288) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@othmankh Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=pivotal/postfacto&pullRequestId=288)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=pivotal/postfacto&pullRequestId=288) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@othmankh Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=pivotal/postfacto&pullRequestId=288)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=pivotal/postfacto&pullRequestId=288) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@othmankh Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=pivotal/postfacto&pullRequestId=288)!

File: api/app/models/retro.rb [hunk 1]
```diff
@@ -97,4 +115,16 @@ def generate_slug
 
     slug
   end
+
+  def should_recompute_join_token?
+    magic_link_enabled? && :will_save_change_to_encrypted_password?
```
Comment (butzopower) [inline api/app/models/retro.rb:120]:
```suggestion
    magic_link_enabled? && will_save_change_to_encrypted_password?
```

The current bit works as is, but as a weird side effect that `false && :will_save_change_to_encrypted_password?` returns a symbol (cuz of default return in ruby) that then gets evaluated in the lifecycle.  Probably better to just call the method directly so you are comparing two booleans and returning a boolean.

File: api/app/models/retro.rb [hunk 2]
```diff
@@ -64,6 +66,22 @@ def validate_login?(val)
...
+    val == join_token
+  end
+
+  def is_magic_link_enabled=(val)
+    if ActiveModel::Type::Boolean.new.cast(val)
+      recompute_join_token
+    else
+      clear_join_token
+    end
+  end
```
Comment (butzopower) [inline api/app/models/retro.rb:79]: If the settings for the retro are updated, but the magic link param isn't touched, does this still cause it to recompute the join token?  Is that desired?

Comment (j-lea) [reply api/app/models/retro.rb:79]: Good catch. Yes it does, but it is not desired. Will fix

**Comment by butzopower** [comment]
There's potentially a way to refactor this in the future to how Devise and other libraries use a "meta"-field like `:password` to cause it to generate something else in certain scenarios, thus allowing you to use out of the box form stuff and not require as much custom controller code.  As is though, looks great, will merge after merge conflict resolved passes CI.

**Pull request merged**